### ISSUE TITLE
[BUG] Chromium fails to download

### ISSUE THREAD

**Issue by mchr3k** [microsoft/playwright#28329]
### System info
- Playwright Version: 1.41.0-alpha-nov-24-2023
- Operating System: macOS 14.0
- Browser: Chromium
- Other info:

**Steps**
- `npm install playwright@next`
- `npx playwright install`

**Expected**

Browsers get installed.

**Actual**

```
% npx playwright install        
Removing unused browser at /Users/martin/Library/Caches/ms-playwright/chromium-1041
Downloading Chromium 120.0.6099.35 (playwright build v1092) from https://playwright.azureedge.net/builds/chromium/1092/chromium-mac-arm64.zip
131 MiB [=================== ] 96% 2.4sError: Download failed: server closed connection. URL: https://playwright.azureedge.net/builds/chromium/1092/chromium-mac-arm64.zip
    at IncomingMessage.<anonymous> (/Users/martin/Documents/GitHub/frontends/node_modules/playwright/node_modules/playwright-core/lib/server/registry/oopDownloadBrowserMain.js:89:22)
    at IncomingMessage.emit (node:events:517:28)
    at emitCloseNT (node:internal/streams/destroy:132:10)
    at process.processTicksAndRejections (node:internal/process/task_queues:81:21)
Downloading Chromium 120.0.6099.35 (playwright build v1092) from https://playwright-akamai.azureedge.net/builds/chromium/1092/chromium-mac-arm64.zip
131 MiB [=================== ] 95% 2.3sError: Download failed: server closed connection. URL: https://playwright-akamai.azureedge.net/builds/chromium/1092/chromium-mac-arm64.zip
    at IncomingMessage.<anonymous> (/Users/martin/Documents/GitHub/frontends/node_modules/playwright/node_modules/playwright-core/lib/server/registry/oopDownloadBrowserMain.js:89:22)
    at IncomingMessage.emit (node:events:517:28)
    at emitCloseNT (node:internal/streams/destroy:132:10)
    at process.processTicksAndRejections (node:internal/process/task_queues:81:21)
Downloading Chromium 120.0.6099.35 (playwright build v1092) from https://playwright-verizon.azureedge.net/builds/chromium/1092/chromium-mac-arm64.zip
131 MiB [====================] 97% 1.2sError: Download failed: server closed connection. URL: https://playwright-verizon.azureedge.net/builds/chromium/1092/chromium-mac-arm64.zip
    at IncomingMessage.<anonymous> (/Users/martin/Documents/GitHub/frontends/node_modules/playwright/node_modules/playwright-core/lib/server/registry/oopDownloadBrowserMain.js:89:22)
    at IncomingMessage.emit (node:events:517:28)
    at emitCloseNT (node:internal/streams/destroy:132:10)
    at process.processTicksAndRejections (node:internal/process/task_queues:81:21)
Failed to install browsers
Error: Failed to download Chromium 120.0.6099.35 (playwright build v1092), caused by
Error: Download failure, code=1
    at ChildProcess.<anonymous> (/Users/martin/Documents/GitHub/frontends/node_modules/playwright/node_modules/playwright-core/lib/server/registry/browserFetcher.js:91:16)
    at ChildProcess.emit (node:events:517:28)
    at ChildProcess._handle.onexit (node:internal/child_process:292:12)
```

This looks related to https://github.com/microsoft/playwright/issues/28189. 

I can successfully download and unzip https://playwright.azureedge.net/builds/chromium/1091/chromium-mac-arm64.zip from Firefox, but it reliably fails to download from `npx playwright install`

**Comment by mchr3k** [comment]
Actually I tried to open the file which was downloaded by Firefox and it unzipped successfully but when I tried to launch Chromium I get an error message "“Chromium” is damaged and can’t be opened. You should move it to the Bin.". This suggests to me that the file on the CDN is genuinely broken.

**Comment by mchr3k** [comment]
"“Chromium” is damaged and can’t be opened. You should move it to the Bin." turns out to be a known issue and I've fixed it by using `brew install chromium` followed by running `xattr -rc /Applications/Chromium.app` as per https://chromium.googlesource.com/chromium/src.git/+/main/docs/mac_arm64.md. I was then able to copy the Chromium binary into place: `cp -r /Applications/Chromium.app /Users/martin/Library/Caches/ms-playwright/chromium-1091/chrome-mac/`.

This has unblocked me for now, but I still can't get playwright to handle downloading Chromium.

**Comment by kachkaev** [comment]
I faced the same issue and shared some thoughts in https://github.com/microsoft/playwright/issues/22878#issuecomment-1826406962.

**TL;DR:** As a workaround, it might be worth trying a different internet connection before anything else.

**Comment by kachkaev** [comment]
I upgraded from `1.40.0` to `1.41.0-alpha-nov-25-2023` and got the same CLI output as shared by @mchr3k above. Before https://github.com/microsoft/playwright/pull/28278, `playwright install` was hanging forever.

@mchr3k you seem to be based in the UK too, so there is some change that we have the same ISP. Can you try a different connection, e.g. tethering or a VPN?

**Comment by mxschmitt** [comment]
Does any of you have a special anti-virus on your system? Since this happened on all three CDNs suggests that the CDN itself behaved correctly, so my guess is that e.g. your system detected it as malicious. But then the point that a different VPN made it works makes me to question this theory.

Could be indeed a ISP issue, which very rarely occurs like in your case and Playwright didn't handle correctly (pre next).

https://github.com/microsoft/playwright/issues/22878#issuecomment-1826406962 great analysis! I like the idea of `response.on('error', ` and then rejecting the promise, sounds like a bit better to what we did in https://github.com/microsoft/playwright/pull/28278. Happy to accept a PR for that, since the tests should continue to pass which we added there as well.

**Comment by kachkaev** [comment]
> Does any of you have a special anti-virus on your system? Since this happened on all three CDNs suggests that the CDN itself behaved correctly, so my guess is that e.g. your system detected it as malicious. But then the point that a different VPN made it works makes me to question this theory.

I’m on macOS, can’t recall any anti-virus software installed. What’s interesting is that even when I am connected without a VPN, I can download and extract Chromium via browsers (i.e. using a different User Agent header). So there must be something really odd happening on the TCP level. I can observe `ECONNRESET` only when using the CLI.

> Happy to accept a PR for that, since the tests should continue to pass which we added there as well.

https://github.com/microsoft/playwright/pull/28344

**Comment by mxschmitt** [comment]
To summarise: <= 1.40 we were hanging. On `main` we don't hang anymore. We fail instead on all CDNs. This does not fix the issue but makes it clearer to the user that the file can't be downloaded.

In https://github.com/microsoft/playwright/pull/28344 we'll improve the way we throw instead of hanging.

**Comment by kachkaev** [comment]
Haha I’m also on Three UK network. Must be something with them then! Thanks for confirming that downloading works for you from EE, similarly to when I enable a VPN on top of Three.

**Comment by mxschmitt** [comment]
Closing as per above then and the enhancement in https://github.com/microsoft/playwright/pull/28344 has landed.

**Comment by Hainesy** [comment]
I am experiencing this and completely blocked by it. I am also on the Three 5G network as my main wifi, and my mobile provider is also Three (I've tried tethering and it didn't work) so I have no alternative provider to try.

**Comment by kachkaev** [comment]
@Hainesy try downloading Playwright via some VPN. If you don’t have a paid one, you can try https://1.1.1.1 by CloudFlare. You don’t need to keep VPN on after downloading binaries.

**Comment by Hainesy** [comment]
@kachkaev I have already tried whilst connected to Cloudflare WARP, with no difference.

I can download the Chromium file directly that fails during the install without problem. But there doesn't appear to be a way to manually get Playwright to recognise it's already downloaded, or at least I haven't figured it out.

**Comment by kachkaev** [comment]
That’s odd. When I was on Three 5G wifi, connecting to a VPN (e.g. CloudFlare Wrap) allowed me to download binaries via `npx playwright install` / `pnpm exec playwright install`.

**Comment by Hainesy** [comment]
I've only been trying to install via `pnpm i`. Running as you say just errors with 'playwright: command not found'. What am I missing?

**Comment by kachkaev** [comment]
If I remember correctly, I removed `playwright` from `package.json` deps and only kept `@playwright/test`. Then I could run `pnpm exec playwright install` before running `pnpm test:system` (→ `playwright test`).

If `playwright` is not a direct dependency, it does not need to be downloaded on `pnpm install`. This makes CI faster.

**Comment by mchr3k** [comment]
@Hainesy - before I worked out that I could just tether to work around my issue, I also managed to get it working by downloading Chrome manually. This is what I did:

---

Init playwright in a package:

```bash
% PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=true npm init playwright@latest
Getting started with writing end-to-end tests with Playwright:
Initializing project in '.'
✔ Where to put your end-to-end tests? · e2e
✔ Add a GitHub Actions workflow? (y/N) · true
✔ Install Playwright browsers (can be done manually via 'npx playwright install')? (Y/n) · false
```

Manually download Chromium and mark it as safe to run as per https://chromium.googlesource.com/chromium/src.git/+/main/docs/mac_arm64.md:

```bash
brew install chromium
xattr -rc /Applications/Chromium.app
```

Manually copy Chromium into the location used by playwright:

```bash
mkdir -p /Users/martinharerobertson/Library/Caches/ms-playwright/chromium-1091/chrome-mac/
cp -r /Applications/Chromium.app /Users/martinharerobertson/Library/Caches/ms-playwright/chromium-1091/chrome-mac/
```

Run playwright tests: `npx playwright test`

---

Hopefully this helps you and anyone else who gets stuck on this.

**Comment by mxschmitt** [comment]
npx playwright install --dry-run

this should also give hints on how to workaround it.

**Comment by david-bezero** [comment]
I also have Three 5G internet and see the same issue. The workaround of downloading it manually fixes it for me too (not sure why it should be different from what it's doing automatically though?)

To make it a bit easier, I wrote a script:

```sh
#!/bin/sh
set -e

INFO="$(yarn playwright install chromium --with-deps --dry-run)"
LOCN="$(echo "$INFO" | grep 'Install location' | sed 's/^[^:]*: *//')"
SRC="$(echo "$INFO" | grep 'Download url' | sed 's/^[^:]*: *//')"

echo "Downloading from $SRC to $LOCN"

mkdir -p "$LOCN"

rm "$LOCN/temp-chrome.zip" 2>/dev/null || true
curl "$SRC" > "$LOCN/temp-chrome.zip"
unzip "$LOCN/temp-chrome.zip" -d "$LOCN"
rm "$LOCN/temp-chrome.zip"
```

(replace `yarn` with `npx` if using NPM as your package manager)

**Comment by shaileshrahinj** [comment]
Thanks @[mchr3k](https://github.com/mchr3k)
, I am also using the Three's internet (birmingham, UK). Faced the exact same issue. But switching over to Lebera(phone n/w), was able to install. :)

**Comment by mattrussell-sonocent** [comment]
FWIW, I was pairing with a colleague today who also uses Three 5G in the UK, and we hit this same problem (connection being closed at ~95%), at all 3 locations it tried. The only workaround we could find was to switch to a mobile data network to complete the download. Downloading the files via `curl` didn't hit any problems either. Very odd!

I recommended my colleague report the problem to their ISP, but in honesty, I doubt think there's any tractable way to get a message past first-line support in these situations...

**Comment by LuaKT** [comment]
Have the same issue on Three, I dug into it a bit and the issue is fixed if you enable keepalive on the [HttpsHappyEyeballsAgent](https://github.com/microsoft/playwright/blob/95fc2b8a8b449560e53999c866e9602f4bca27f1/packages/playwright-core/src/utils/happy-eyeballs.ts#L39) to match the default [https.globalAgent](https://nodejs.org/api/https.html#httpsglobalagent). I'm not sure what causes it on Three only or why this fixes it though.

Are there any downsides to enabling keepalive here anyway?

**Comment by mxschmitt** [comment]
@LuaKT awesome discovery! Do you mind sending a PR for this? Otherwise we can do it as well.

**Comment by Eden13300** [comment]
so for running docker how to change the keepalive?

**Comment by tribet84** [comment]
This is common in corporate environments with ssl inspection, the common approach is to tell node to accept the self-signed certificate

`NODE_TLS_REJECT_UNAUTHORIZED=0 npx playwright install`
